exp/api/v1alpha4/azuremanagedmachinepool_webhook.go

```diff
@@ -0,0 +1,179 @@
+	"sigs.k8s.io/controller-runtime/pkg/webhook"
+)
+
+// log is for logging in this package.
+var azuremanagedmachinepoollog = logf.Log.WithName("azuremanagedmachinepool-resource")
+
+var webhookClient client.Client
+
+// SetupWebhookWithManager sets up and registers the webhook with the manager.
+func (r *AzureManagedMachinePool) SetupWebhookWithManager(mgr ctrl.Manager) error {
+	return ctrl.NewWebhookManagedBy(mgr).
+		For(r).
+		Complete()
+}
+
+// NewAzureManagedMachinePoolWebhook returns a new AzureManagedMachinePool instance.
+func NewAzureManagedMachinePoolWebhook(client client.Client) *AzureManagedMachinePool {
+	webhookClient = client
+	return &AzureManagedMachinePool{}
+}
+
+//+kubebuilder:webhook:path=/mutate-infrastructure-cluster-x-k8s-io-v1alpha4-azuremanagedmachinepool,mutating=true,failurePolicy=fail,groups=infrastructure.cluster.x-k8s.io,resources=azuremanagedmachinepools,verbs=create;update,versions=v1alpha4,name=default.azuremanagedmachinepools.infrastructure.cluster.x-k8s.io,sideEffects=None,admissionReviewVersions=v1beta1
+
+var _ webhook.Defaulter = &AzureManagedMachinePool{}
+
+// Default implements webhook.Defaulter so a webhook will be registered for the type.
+func (r *AzureManagedMachinePool) Default() {
+	azuremanagedmachinepoollog.Info("default", "name", r.Name)
+
+	if r.Labels == nil {
+		r.Labels = make(map[string]string)
+	}
+	r.Labels[LabelAgentPoolMode] = r.Spec.Mode
+}
+
+//+kubebuilder:webhook:verbs=update;delete,path=/validate-infrastructure-cluster-x-k8s-io-v1alpha4-azuremanagedmachinepool,mutating=false,failurePolicy=fail,groups=infrastructure.cluster.x-k8s.io,resources=azuremanagedmachinepools,versions=v1alpha4,name=validation.azuremanagedmachinepools.infrastructure.cluster.x-k8s.io,sideEffects=None,admissionReviewVersions=v1beta1
+
+var _ webhook.Validator = &AzureManagedMachinePool{}
+
+// ValidateCreate implements webhook.Validator so a webhook will be registered for the type.
+func (r *AzureManagedMachinePool) ValidateCreate() error {
+	azuremanagedmachinepoollog.Info("validate create", "name", r.Name)
+	return nil
+}
+
+// ValidateUpdate implements webhook.Validator so a webhook will be registered for the type.
+func (r *AzureManagedMachinePool) ValidateUpdate(oldRaw runtime.Object) error {
+	old := oldRaw.(*AzureManagedMachinePool)
+	var allErrs field.ErrorList
+
+	if r.Spec.SKU != old.Spec.SKU {
+		allErrs = append(allErrs,
+			field.Invalid(
+				field.NewPath("Spec", "SKU"),
+				r.Spec.SKU,
+				"field is immutable"))
+	}
+
+	if old.Spec.OSDiskSizeGB != nil {
+		// Prevent OSDiskSizeGB modification if it was already set to some value
+		if r.Spec.OSDiskSizeGB == nil {
+			// unsetting the field is not allowed
+			allErrs = append(allErrs,
+				field.Invalid(
+					field.NewPath("Spec", "OSDiskSizeGB"),
+					r.Spec.OSDiskSizeGB,
+					"field is immutable, unsetting is not allowed"))
+		} else if *r.Spec.OSDiskSizeGB != *old.Spec.OSDiskSizeGB {
+			// changing the field is not allowed
```

Comment:
I think os disk size is mutable as long as you use managed disks, but I could be mistaken.

Reply:
will check and get back on this.